### DISCUSSION TITLE
New cluster with old config

### DISCUSSION THREAD

**Discussion by emilioziniades** [rook/rook#14695] [Q&A]
Hi there.

Firstly, thank you for making rook, it's a brilliant tool and I have set it up with little difficulty.

I wanted to ask: is it possible for a new kubernetes cluster to adopt an existing rook-ceph cluster?

I have a [host storage cluster](https://rook.io/docs/rook/latest-release/CRDs/Cluster/host-cluster/). Also, I have read [Backing up and restoring a cluster based on PVCs into a new Kubernetes cluster](https://rook.io/docs/rook/latest-release/Troubleshooting/disaster-recovery/#backing-up-and-restoring-a-cluster-based-on-pvcs-into-a-new-kubernetes-cluster), but this is not the exact same setup as mine, because the ceph config is stored on the hosts directly.

In my case, the config and raw OSD devices can be easily moved to new hosts. If I spin up a new kubernetes cluster on these new hosts and re-attach the config disk and the OSD devices before recreating the rook-cluster (using helm), will rook-ceph realize there is existing config and use that config, or do I have to do something extra (such as in the linked article) to do the migration?

I'm happy to provide more details if necessary.

**Comment by travisn** [comment]
Is this a scenario that you want to automate? Or it's a one-off that you hope to get working for one of your clusters? The rook documentation just assumes it's a disaster recovery scenario, so the steps are very manual.

This is expected to be possible, assuming the critical ceph metadata is restored in the new cluster, such as admin keyring, same mon endpoints, etc, which should be all listed in that doc you linked. If you find something isn't working, we would certainly like to find the solution and update the doc.

**Reply by emilioziniades** [reply]
Yeah it's something I'd like to automate. 

Ok cool, just to clarify, when you say "critical metadata" are you referring to the k8s configmaps and secrets?

I will experiment and let you know if it worked.

**Comment by emilioziniades** [comment]
@travisn so, I was able to follow those steps in the disaster recovery docs and adopt the existing rook-ceph cluster into a new kubernetes cluster. The existing ceph filesystem was there. However, the PVs were gone because it is a fresh kubernetes cluster. What I really wanted to acheive was a backup and restore of all the PVs/PVCs between an old k8s cluster and a new k8s cluster. Do you have any suggestions for how do to that? I've seen Velero being mentioned a lot on here.

**Reply by travisn** [reply]
Velero would backup/restore the data, but it also involves moving the data out and back into the cluster, which is not your intention. In this scenario, the data is already in the restored cluster, but you just need to re-create the PVs and PVCs so the applications can consume the same volumes.

You may be able to use [Static PVs](https://github.com/ceph/ceph-csi/blob/49926446f998da465f949e97f1acd86e6837eb30/docs/static-pvc.md#static-pvc-with-ceph-csi) for the rbd and cephfs volumes. If you first export the PVC and PV metadata from the old cluster, you should be able to re-generate static PVs from the original PVs. @Madhu-1 @Rakshith-R Other ideas on how to restore PVs from a rebuilt cluster?

**Comment by emilioziniades** [comment]
I will document the steps I took incase it is useful for anyone else.

Assuming you have (1) a running kubernetes cluster ("old k8s cluster"), (2) are connected via `kubectl` to the old k8s cluster and (3) the rook-ceph cluster has been set up in the `rook-ceph` namespace:

1. Back up the necessary services, secrets and configmaps:
```
mkdir -p backup

# backup mon services
kubectl get service -n rook-ceph |
    grep "rook-ceph-mon" |
    awk '{print $1}' |
    xargs -I {} bash -c "kubectl get service -n rook-ceph {} -o yaml > backup/{}.yaml"

# backup mon-endpoints configmap
kubectl get configmap -n rook-ceph |
    grep "mon-endpoints" |
    awk '{print $1}' |
    xargs -I {} bash -c "kubectl get configmap -n rook-ceph {} -o yaml > backup/{}.yaml"

# backup fsid and keyring secrets
kubectl get secret -n rook-ceph |
    grep "keyring\|rook-ceph-mon" |
    awk '{print $1}' |
    xargs -I {} bash -c "kubectl get secret -n rook-ceph {} -o yaml > backup/{}.yaml"
```
2. Tear down the old k8s cluster, BUT do not delete the ceph config (`dataDirHostPath`) directory on the hosts, and do not reformat the raw devices used by ceph.
3. Create new k8s cluster, ensuring that the ceph config (`dataDirHostPath`) is present, as well as the raw devices. (In my case, I simply destroyed the kubernetes cluster and recreated it on the same hosts)
4. Deploy the rook operator
5. Restore the backed up data above: `kubectl apply -f backup`
6. Deploy the rook cluster

Now, you have a new kubernetes cluster which has adopted the existing ceph cluster.

This works, but a few provisos:
- The new k8s cluster must have the _exact_ same topology as before. There should be the same nodes at the same IP addresses. The CNI must be set up exactly as before. This is important because the `rook-ceph-mon-endpoints` configmap and the mon services from the old k8s cluster expect the same nodes to be there.
- The rook cluster must be exactly the same as before. This is required because we manually recreate the mon services from a backup.
- You will not have access to your PVs or PVCs from the old cluster. The files will still be there in the ceph cluster, but you have no way of accessing them unless you use [direct tools](https://rook.io/docs/rook/latest-release/Troubleshooting/direct-tools/).
